File: 3Sum.cpp
/*

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
	[-1, 0, 1],
	[-1, -1, 2]
]

*/


#include <iostream>
#include <vector>
#include <numeric>
#include <functional>
#include <algorithm>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

using namespace std;

vector<vector<int> > threeSum(vector<int> &nums) {

	vector<vector<int> > res;
	std::sort(nums.begin(), nums.end());
	int target, front, back, sum;

	for (int i = 0; i < nums.size(); i++) {
		target = -nums[i];
		front = i + 1;
		back = nums.size() - 1;

		while (front < back) {
			sum = nums[front] + nums[back];
			// Finding answer which start from number num[i]
			if (sum < target)
				front++;
			else if (sum > target)
				back--;
			else {
				vector<int> triplet(3, 0);
				triplet[0] = nums[i];
				triplet[1] = nums[front];
				triplet[2] = nums[back];
				res.push_back(triplet);
				// Processing duplicates of Number 2
				// Rolling the front pointer to the next different number forwards
				while (front < back && nums[front] == triplet[1]) front++;
				// Processing duplicates of Number 3
				// Rolling the back pointer to the next different number backwards
				while (front < back && nums[back] == triplet[2]) back--;
			}
		}

		// Processing duplicates of Number 1
		while (i + 1 < nums.size() && nums[i + 1] == nums[i]) i++;

	}

	return res;

}


int main() {

	vector<int> input_vec = { -1,0,1,2,-1,-4,-3,3,2,5,-1 };
	vector<vector<int>> output_vec = threeSum(input_vec);

	cout << "ANSWER IS " << endl;
	for (auto i : output_vec) {
		for (auto j : i) {
			cout << j << ", ";
		}
		cout << endl;
	}

	system("pause");
	return 0;
	
}

File: py/3Sum.py
'''

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? 
Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],
A solution set is:
[
	[-1, 0, 1],
	[-1, -1, 2]
]

'''


class Solution:
    '''
    Find two number sums == target
    Reduce the problem to 2Sum
    '''
    def threeSum(self, nums):
        results = []
        nums = sorted(nums)
        target, front, back, total = 0,0,0,0
        
        for ii in range(len(nums)):
            target = -nums[ii]      # Find two number sums == target
            front = ii + 1
            back = len(nums) - 1
            
            while front < back:
                total = nums[front] + nums[back]
                # total + target < 0
                if total < target:
                    front += 1      # left goes right
                # total + target > 0
                elif total > target:
                    back -= 1       # right goes less
                else:
                    result = [nums[ii], nums[front], nums[back]]
                    results.append(result)
                    print(ii, front, back)
                    while front < back and nums[front] == result[1]:
                        front += 1
                    while front < back and nums[back] == result[2]:
                        back -= 1 
                    # <QUITE ODD HERE>
                    break
            while ii+1 < len(nums) and nums[ii+1] == nums[ii]:
                ii += 1
        return results


if __name__ == '__main__':
    solu = Solution()
    inputs = [0,0,0]
    print(inputs, solu.threeSum(inputs))